### PR TITLE
Initial sparc64-linux bringup

### PR DESCRIPTION
**Pull request by koachan**
This is to track ongoing effort to bring Zig to Linux on sparc64 as requested in #4931.

Milestones:
- [x] Syscall interface
- [x] os.bits
- [x] C ABI
- [x] clone(2) implementation
- [x] TLS
- [ ] Compiles "hello world" program
- [ ] Compiles C program with --c-source
- [ ] Passes test suite

### CODE REVIEW

**Comment by LemonBoy** [comment]
> clone(2) implementation

In case you need one, here's mine:

```zig
            // __clone(func, stack, flags, arg, ptid, tls, ctid)
            //           i0,    i1,    i2,  i3,   i4,  i5,   sp

            // syscall(SYS_clone, flags, stack, ptid, tls, ctid)
            //                g1     o0,    o1,   o2,  o3,   o4
            asm volatile (
                \\ save %%sp, -192, %%sp
                \\ # Save the func pointer and the arg pointer
                \\ mov %%i0, %%g2
                \\ mov %%i3, %%g3
                \\ # Shuffle the arguments
                \\ mov 217, %%g1
                \\ mov %%i2, %%o0
                \\ sub %%i1, 2047, %%o1
                \\ mov %%i4, %%o2
                \\ mov %%i5, %%o3
                \\ ldx [%%fp + 192 - 2*8 + 2047], %%o4
                \\ t 0x6d
                \\ bcs,pn %%xcc, 2f
                \\ nop
                \\ # sparc64 returns the child pid in o0 and a flag telling
                \\ # whether the process is the child in o1
                \\ brnz %%o1, 1f
                \\ nop
                \\ # This is the parent process, return the child pid
                \\ mov %%o0, %%i0
                \\ ret
                \\ restore
                \\1:
                \\ # This is the child process
                \\ mov %%g0, %%fp
                \\ call %%g2
                \\ mov %%g3, %%o0
                \\ # Exit
                \\ mov 1, %%g1
                \\ t 0x6d
                \\2:
                \\ # The syscall failed
                \\ sub %%g0, %%o0, %%i0
                \\ ret
                \\ restore
            );
```

Oh, there's one more item in the todo list: a stack-walking routine is needed. Good luck with that.

**Comment by koachan** [comment]
> > clone(2) implementation
> 
> In case you need one, here's mine:
Okay, I'll use that, thanks a lot!
 
> Oh, there's one more item in the todo list: a stack-walking routine is needed. Good luck with that.
Wait, so just putting the starting pointer in lib/std/start.zig isn't enough?

File: lib/std/os/linux.zig [hunk 1]
```diff
@@ -373,7 +374,7 @@ pub fn faccessat(dirfd: i32, path: [*:0]const u8, mode: u32, flags: u32) usize {
 }
 
 pub fn pipe(fd: *[2]i32) usize {
-    if (comptime builtin.arch.isMIPS()) {
+    if (comptime builtin.arch.isMIPS() || comptime builtin.arch.isSPARC()) {
```
Comment (daurnimator) [inline lib/std/os/linux.zig:377]: I'd put the comptime outside of the parenthesis: `comptime (builtin.arch.isMIPS() || builtin.arch.isSPARC())`

Comment (koachan) [reply lib/std/os/linux.zig]: I've moved it out, thanks for the suggestion!

**Comment by andrewrk** [comment]
> > Oh, there's one more item in the todo list: a stack-walking routine is needed. Good luck with that.
> 
> Wait, so just putting the starting pointer in lib/std/start.zig isn't enough?

I believe @LemonBoy is referring to this:

https://github.com/ziglang/zig/blob/master/lib/std/special/compiler_rt/stack_probe.zig

It's OK to stub it out and we can open an issue for it to track it. It doesn't need to block the rest of your work here.

**Comment by LemonBoy** [comment]
> I believe @LemonBoy is referring to this:

Nope, stack probes are (sadly) only available on x86/x64. What I was referring to is the code in `debug.zig` that walks the stack frames. Sparc has a few quirks to take into account such as the stack being biased by 2047 bytes, the presence of the 192 byte area (IIRC that's only for non-leaf functions) to spill the window contents, the fact you need to `flushw` the windows on stack before trying to walk it and a few other details I'm probably forgetting.

**Comment by koachan** [comment]
Hmm, so, I have two questions:
- I just found out that lld only gotten support for sparc64 in version 11, and it seems that the current master doesn't have support for LLVM 11 yet. Should I rebase on top of the llvm11 branch, or should I wait for LLVM 11 support to be merged to master?
- Is it fine if I stub the stack-walking function and work on it later on?

**Comment by andrewrk** [comment]
> * I just found out that lld only gotten support for sparc64 in version 11, and it seems that the current master doesn't have support for LLVM 11 yet. Should I rebase on top of the llvm11 branch, or should I wait for LLVM 11 support to be merged to master?

llvm11 branch is merged into master now, so you should be unblocked on this front :)

> * Is it fine if I stub the stack-walking function and work on it later on?

Yes it is fine. Better to send smaller patches that can be merged independently so that we don't lose all your hard work.


I'm going to close this pull request due to inactivity, merge conflicts, and tests not passing. However I hope that you open a new one and finish! This was looking pretty good. In case you don't have time to do this, someone else can finish your work by finding a link to this PR in the issue #4931. Again please try to take advantage of zig's lazy compilation. It should be possible to do this work much more incrementally so that your work does not have to reach a particular milestone before getting merged.

**Comment by koachan** [comment]
Thanks! I'll rebase it on the new master and try to get it to build simple programs, then I'll open a new PR.

**Pull request closed without merging**